### PR TITLE
fix: add missing KZG point evaluation precompile

### PR DESCRIPTION
**Pull request by Wodann**
In the process of fixing this, I also updated all Rust dependencies to ensure that we could use the latest version of `revm` (v8.0.0)

**Depends on**
- [x] https://github.com/bluealloy/revm/pull/1276
- [x] https://github.com/bluealloy/revm/pull/1284

### CODE REVIEW

File: crates/edr_eth/src/transaction/signed.rs
Comment (agostbiro) [inline crates/edr_eth/src/transaction/signed.rs:52]: Nice to have: I'm wondering if we can have a single implementation for the logic in the branches and have some basic unit tests for edge cases?

File: crates/edr_evm/src/mempool.rs
Comment (agostbiro) [inline crates/edr_evm/src/mempool.rs:28]: Question: I'm wondering what's the reason to change to `shift_remove`? Based on skimming the docs `remove` is `O(1)` while `shift_remove` is `O(n)`. Can that be a problem? Are we potentially calling this in a loop making it `O(n^2)`?

File: crates/edr_provider/src/data.rs
Comment (agostbiro) [inline crates/edr_provider/src/data.rs:3746]: Change request: could you add a comment to the code explaining how this test is different from the `mainnet_cancun` test?

File: crates/edr_provider/src/test_utils.rs
Comment (agostbiro) [inline crates/edr_provider/src/test_utils.rs:194]: Tentative change request: I'm wondering if it would be better to use normal assert here in case the tests are ran in release mode.

**Review by agostbiro — CHANGES_REQUESTED** [review]

**Comment by agostbiro** [comment]
Change request: it looks like there is a new transitive dependency that doesn't build on Windows.

File: crates/edr_evm/src/mempool.rs
Comment (fvictorio) [reply crates/edr_evm/src/mempool.rs:28]: I guess it's because the data structure, [according to its docs](https://crates.io/crates/indexmap):

> Preserves insertion order as long as you don't call .remove(), .swap_remove(), or other methods that explicitly change order. The alternate .shift_remove() does preserve relative order.

I wonder if we have a test for the scenario where:

1) Automining is disabled
2) Mempool is used in FIFO mode
3) 5 transactions are added to the mempool
4) 3rd transaction is removed
5) The rest of the txs preserve the order

Comment (Wodann) [reply crates/edr_evm/src/mempool.rs:28]: > I'm wondering what's the reason to change to shift_remove?

`remove` was deprecated because it was silently reordering. Instead they added two explicit methods that signal reordering vs order-preserving

File: crates/edr_provider/src/test_utils.rs
Comment (Wodann) [reply crates/edr_provider/src/test_utils.rs:194]: This is explicitly using debug asserts such that they are not run in release mode.

For this reason tests should never be run in release mode, but instead you should enable optimisations for the test profile

File: crates/edr_eth/Cargo.toml
Comment (Wodann) [reply crates/edr_eth/Cargo.toml:23]: We upstreamed 2 changes to `revm` that:
1. was merged but not released yet
2. was not merged yet

File: crates/edr_eth/src/transaction/signed.rs
Comment (Wodann) [reply crates/edr_eth/src/transaction/signed.rs:52]: > I'm wondering if we can have a single implementation for the logic in the branches

No, these are two different types: EIP-1559 transaction and EIP-4844 transaction, so it's impossible to merge the two cases

Comment (agostbiro) [reply crates/edr_eth/src/transaction/signed.rs:52]: > No, these are two different types: EIP-1559 transaction and EIP-4844 transaction, so it's impossible to merge the two cases

A macro could work, but probably an overkill

> Did you have anything in mind?

I'm not sure what would happen if `tx.max_fee_per_gas` is less than `block_base_fee`

(none of these are merge blockers for me)

File: crates/edr_evm/src/mempool.rs
Comment (agostbiro) [reply crates/edr_evm/src/mempool.rs:28]: Thanks, I checked now and it seems that we're using `remove_caller` in a loop causing worst case `O(n^2)` removal:

https://github.com/NomicFoundation/hardhat/blob/69e340761ce52812fe720689bcec972345fc8df7/crates/edr_evm/src/miner.rs#L134-L138

We should turn that into worst case linear before merging this imo.

Comment (Wodann) [reply crates/edr_evm/src/mempool.rs:28]: > Thanks, I checked now and it seems that we're using `remove_caller` in a loop causing worst case `O(n^2)` removal:
> 
> https://github.com/NomicFoundation/hardhat/blob/69e340761ce52812fe720689bcec972345fc8df7/crates/edr_evm/src/miner.rs#L134-L138
> 
> We should turn that into worst case linear before merging this imo.

The order needs to be preserved, as Franco pointed out, so the entries have to be shifted.

Comment (fvictorio) [reply crates/edr_evm/src/mempool.rs:28]: Technically we only need to preserve ordering when the mempool is in FIFO mode. But I think it's fine for this to be quadratic, N here is the number of transactions in the mempool, which is very unlikely to be a large number.

**Review by fvictorio — APPROVED** [review]

File: crates/edr_evm/src/mempool.rs
Comment (agostbiro) [reply crates/edr_evm/src/mempool.rs:28]: Ok created tech debt issue based on in-person discussion: https://github.com/NomicFoundation/edr/issues/382

File: crates/edr_provider/src/test_utils.rs
Comment (Wodann) [reply crates/edr_provider/src/test_utils.rs:194]: We discussed in person, but for posterity: The `debug_assert`s are there for debugging purposes. They provide additional information for when the `assert` fails, specifically in the case of unequal `receipt_root`s.

File: crates/edr_provider/src/data.rs
Comment (Wodann) [reply crates/edr_provider/src/data.rs:3748]: Addressed in https://github.com/NomicFoundation/hardhat/pull/5100/commits/03796a3f34646b5fb42db7e342149f9c82fe76e5

**Review by agostbiro — APPROVED** [review]
Thanks, LGTM assuming CI passes.

**Pull request merged**
